demos/guide/quic-server-non-block.c

```diff
@@ -0,0 +1,478 @@
+         * stream reset - or some failure occurred on the underlying
+         * connection.
+         */
+        switch (SSL_get_stream_read_state(ssl)) {
+        case SSL_STREAM_STATE_RESET_REMOTE:
+            printf("Stream reset occurred\n");
+            /*
+             * The stream has been reset but the connection is still
+             * healthy.
+             */
+            break;
+
+        case SSL_STREAM_STATE_CONN_CLOSED:
+            printf("Connection closed\n");
+            /* Connection is already closed. */
+            break;
+
+        default:
+            printf("Unknown stream failure\n");
+            break;
+        }
+        /*
+         * If the failure is due to a verification error we can get more
+         * information about it from SSL_get_verify_result().
+         */
+        if (SSL_get_verify_result(ssl) != X509_V_OK)
+            printf("Verify error: %s\n",
+                   X509_verify_cert_error_string(SSL_get_verify_result(ssl)));
+        return -1;
+
+    default:
+        return -1;
+    }
+}
+
+/*
+ * Main loop for server to accept QUIC connections.
+ * Echo every request back to the client.
+ */
+static int run_quic_server(SSL_CTX *ctx, int fd)
+{
+    int ok = -1;
+    int ret, eof;
+    SSL *listener, *conn = NULL;
+    unsigned char buf[8192];
+    size_t nread, nwritten, total_read, total_written;
+
+    /* Create a new QUIC listener */
+    if ((listener = SSL_new_listener(ctx, 0)) == NULL)
+        goto err;
+
+    /* Provide the listener with our UDP socket. */
+    if (!SSL_set_fd(listener, fd))
+        goto err;
+
+    /*
+     * Set the listener mode to non-blocking, which is inherited by
+     * child objects.
+     */
+    if (!SSL_set_blocking_mode(listener, 0))
+        goto err;
+
+    /* Begin listening. */
+    if (!SSL_listen(listener))
+        goto err;
+
+    /*
+     * Begin an infinite loop of listening for connections. We will only
+     * exit this loop if we encounter an error.
+     */
+    for (;;) {
+        eof = 0;
+        total_read = 0;
+        total_written = 0;
+
+        /* Pristine error stack for each new connection */
+        ERR_clear_error();
+
+        /* Block while waiting for a client connection */
+        printf("Waiting for connection\n");
+        while ((conn = SSL_accept_connection(listener, 0)) == NULL)
+            wait_for_activity(listener);
+        printf("Accepted new connection\n");
+
+        /* Read from client until the client sends a end of stream packet */
+        while (!eof) {
+            ret = SSL_read_ex(conn, buf, sizeof(buf), &nread);
+            total_read += nread;
+            switch (handle_io_failure(conn, ret)) {
+            case 1:
+                continue; /* Retry */
+            case 0:
+                /* Reached end of stream */
+                if (SSL_has_pending(conn)) {
+                    if (SSL_read_ex(conn, buf, sizeof(buf), &nread) <= 0) {
```

Comment:
this code is bit confusing for me. my confusion comes from fact we repeatedly call `SSL_read_ex()` here to read more data, but we let fresh data to overwrite data in `buf` we keep in `buf` from the previous iteration.